### PR TITLE
feat(core): Allow overriding npm registry for community packages

### PR DESCRIPTION
**Pull request by netroy**
## Summary
This PR adds the option to override npm registry for installing community packages.
This also works as a security fix against a compromised `.npmrc`, by being explicitly using `--registry` in all `npm install` commands.

[Docs PR](https://github.com/n8n-io/n8n-docs/pull/2328)

## Related Linear tickets, Github issues, and Community forum posts

SEC-59

## Review / Merge checklist

- [x] PR title and summary are descriptive
- [x] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included

### CODE REVIEW

**Review by ivov — COMMENTED** [review]
Do you know of an alternate registry with community packages so I can test this?

File: packages/@n8n/config/src/configs/nodes.ts [hunk 1]
```diff
@@ -26,6 +26,10 @@ class CommunityPackagesConfig {
 	@Env('N8N_COMMUNITY_PACKAGES_ENABLED')
 	enabled: boolean = true;
 
+	/** NPM registry to pull community packages from */
+	@Env('N8N_COMMUNITY_PACKAGES_REGISTRY')
+	registry: string = 'https://registry.npmjs.org';
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/@n8n/config/src/configs/nodes.ts:31]: I think we should do this in a separate PR, and update all the config parameters.

**Comment by ivov** [comment]
> > Do you know of an alternate registry with community packages so I can test this?
> 
> We could publish a test package to [the Github NPM registry](https://docs.github.com/en/actions/publishing-packages/publishing-nodejs-packages#publishing-packages-to-github-packages)

I thought this was tested. Or because it's a security fix we need to merge this urgently?

**Comment by netroy** [comment]
> I thought this was tested.

I did test it with a local [verdaccio](https://github.com/verdaccio/verdaccio).

**Comment by ivov** [comment]
Sorry, I'm not familiar with either of those registries, so I'll have to come back later to see how to set them up to test this properly. Let me know if you need this merged anyway.

**Comment by netroy** [comment]
To test this:
1. run `docker run -it --rm -p 4873:4873 verdaccio/verdaccio:nightly-master`
2. set the registry env variable to `http://127.0.0.1:4873`
3. start the app, and install any of the community nodes

everything should work just as it did before, but you can see the proxy calls in the container logs.

**Review by ivov — DISMISSED** [review]
Tested and working. Pending internal discussion.

File: packages/cli/src/constants.ts [hunk 2]
```diff
@@ -90,6 +90,7 @@ export const LICENSE_FEATURES = {
 	PROJECT_ROLE_ADMIN: 'feat:projectRole:admin',
 	PROJECT_ROLE_EDITOR: 'feat:projectRole:editor',
 	PROJECT_ROLE_VIEWER: 'feat:projectRole:viewer',
+	COMMUNITY_NODES_CUSTOM_REGISTRY: 'feat:communityNodes:customRegistry',
```
Comment (ivov) [inline packages/cli/src/constants.ts:93]: I see it's not in the [license server setup](https://github.com/search?q=repo%3An8n-io%2Flicense-management-setup%20communityNodes&type=code) repo yet - we should include it there by the time we release this.

File: packages/cli/src/services/__tests__/communityPackages.service.test.ts [hunk 3]
```diff
@@ -394,6 +397,7 @@ describe('CommunityPackagesService', () => {
 			installedPackage.packageName = mockPackageName();
 
 			mocked(exec).mockImplementation(execMock);
+			license.isCustomNpmRegistryEnabled.mockReturnValue(true);
```
Comment (ivov) [inline packages/cli/src/services/__tests__/communityPackages.service.test.ts:400]: Should we test for no-license-feature case?

File: packages/cli/src/services/communityPackages.service.ts [hunk 4]
```diff
@@ -24,6 +24,9 @@ import type { CommunityPackages } from '@/Interfaces';
 import { LoadNodesAndCredentials } from '@/LoadNodesAndCredentials';
 import { Logger } from '@/Logger';
 import { OrchestrationService } from './orchestration.service';
+import { License } from '@/License';
+
+const DEFAULT_REGISTRY = 'https://registry.npmjs.org';
```
Comment (ivov) [inline packages/cli/src/services/communityPackages.service.ts:29]: Let's update the docs that custom registries require a license?

File: packages/cli/src/services/communityPackages.service.ts [hunk 5]
```diff
@@ -325,13 +324,22 @@ export class CommunityPackagesService {
 		await this.orchestrationService.publish('community-package-uninstall', { packageName });
 	}
 
+	private getNpmRegistry() {
+		let { registry } = this.globalConfig.nodes.communityPackages;
+		if (registry !== DEFAULT_REGISTRY && !this.license.isCustomNpmRegistryEnabled()) {
+			this.logger.warn('Not licensed to use custom NPM registry');
+			registry = DEFAULT_REGISTRY;
```
Comment (ivov) [inline packages/cli/src/services/communityPackages.service.ts:331]: A paying user whose license expires might inadvertently end up installing or updating from `DEFAULT_REGISTRY`.

Should we throw `FeatureNotLicensedError` instead and ask them to remove `N8N_COMMUNITY_PACKAGES_REGISTRY` or renew the license?

File: packages/cli/src/constants.ts (hunk 2 above)
Comment (netroy) [reply packages/cli/src/constants.ts:93]: yeah. was planning to update the product yaml files after checking with David or Cornelius.

Comment (netroy) [reply packages/cli/src/constants.ts:93]: https://github.com/n8n-io/license-management-setup/pull/60

File: packages/cli/src/services/communityPackages.service.ts (hunk 4 above)
Comment (netroy) [reply packages/cli/src/services/communityPackages.service.ts:29]: [will do](https://github.com/n8n-io/n8n-docs/pull/2328#discussion_r1715449176)

**Review by ivov — APPROVED** [review]
💎

**Comment by janober** [comment]
Got released with [`n8n@1.55.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.55.0)

**Pull request merged**